### DISCUSSION TITLE
Filesystem: How to Integrate and Flash PEM and INI Files into STM32 Internal Flash Using Zephyr?

### DISCUSSION THREAD

**Discussion by srikanthpalvai** [zephyrproject-rtos/zephyr#73478] [Q&A]
I am working on a project using an STM32 microcontroller with the Zephyr OS, and I need to integrate and flash a cacert.pem file and a config.ini file into the internal flash memory. My development board does not have an SD card integrated, so I need to use the internal flash memory for this purpose. At runtime, I will load the cacert.pem file for mbedTLS usage and read configurations from the config.ini file.

Are there any specific Zephyr features or best practice example avaibale in this use case ?

**Comment by MaazSk** [comment]
Hi,
Are you able to read and write to internal flash. I am also trying the same

**Comment by nordicjm** [comment]
Have you look at how the sample does it here in the cmake file? https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/net/sockets/http_client

**Comment by MaazSk** [comment]
I don't need to store certificates but I need to write to the flash. And yes I started looking into the examples provided by zephyr

**Reply by nordicjm** [reply]
Then there is nvs which is light weight: https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/subsys/nvs
Or file system support which is heavy: https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/subsys/fs/fs_sample

Though note there are restrictions on usage for both which a lot of st parts have incompatibilities with, you need small uniform flash sector sizes e.g. 4KiB, sector sizes like 16KiB+ will not work with nvs and will not work well with file systems

**Reply by MaazSk** [reply]
Thank you! I was thinking something similar to NVS also is there a way to make it write protect?

**Reply by nordicjm** [reply]
There is not, you can set a flag in your application that prevents your code from calling the write functions, or if your hardware supports locking sectors could use that

**Reply by MaazSk** [reply]
Using nvs with locking sectors right?
